numpyro/contrib/tfp/mcmc.py

```diff
@@ -0,0 +1,194 @@
+    :param callable init_strategy: a per-site initialization function.
+        See :ref:`init_strategy` section for available functions.
+    :param kernel_kwargs: other arguments to be passed to TFP kernel constructor.
+    """
+    kernel_class = None
+
+    def __init__(self, model=None, potential_fn=None, init_strategy=init_to_uniform,
+                 **kernel_kwargs):
+        if not (model is None) ^ (potential_fn is None):
+            raise ValueError('Only one of `model` or `potential_fn` must be specified.')
+        self._model = model
+        self._potential_fn = potential_fn
+        self._kernel_kwargs = kernel_kwargs
+        self._init_strategy = init_strategy
+        # Set on first call to init
+        self._init_fn = None
+        self._postprocess_fn = None
+        self._sample_fn = None
+
+    def _init_state(self, rng_key, model_args, model_kwargs, init_params):
+        if self._model is not None:
+            init_params, potential_fn, postprocess_fn, model_trace = initialize_model(
+                rng_key,
+                self._model,
+                init_strategy=self._init_strategy,
+                dynamic_args=True,
+                model_args=model_args,
+                model_kwargs=model_kwargs)
+            init_params = init_params.z
+            if self._init_fn is None:
+                _, unravel_fn = ravel_pytree(init_params)
+                kernel = self.kernel_class(
+                    _make_log_prob_fn(potential_fn(*model_args, **model_kwargs), unravel_fn),
+                    **self._kernel_kwargs)
+                self._init_fn, self._sample_fn = _extract_kernel_functions(kernel)
+            self._postprocess_fn = postprocess_fn
+        elif self._init_fn is None:
+            _, unravel_fn = ravel_pytree(init_params)
+            kernel = self.kernel_class(
+                _make_log_prob_fn(self._potential_fn, unravel_fn),
+                **self._kernel_kwargs)
+            self._init_fn, self._sample_fn = _extract_kernel_functions(kernel)
+        return init_params
+
+    @property
+    def model(self):
+        return self._model
+
+    @property
+    def sample_field(self):
+        return 'z'
+
+    @property
+    def default_fields(self):
+        return ('z',)
+
+    def get_diagnostics_str(self, state):
+        """
+        Given the current `state`, returns the diagnostics string to
+        be added to progress bar for diagnostics purpose.
+        """
+        return ''
+
+    def init(self, rng_key, num_warmup, init_params=None, model_args=(), model_kwargs={}):
+        # non-vectorized
+        if rng_key.ndim == 1:
+            rng_key, rng_key_init_model = random.split(rng_key)
+        # vectorized
+        else:
+            rng_key, rng_key_init_model = jnp.swapaxes(vmap(random.split)(rng_key), 0, 1)
+        init_params = self._init_state(rng_key_init_model, model_args, model_kwargs, init_params)
+        if self._potential_fn and init_params is None:
```

Comment:
I see that you are using `vmap` for vectorization - are tfp mcmc kernels not auto vectorized?

Reply:
I just took a look at tfp mcmc. It seems there is another utility [sample_chain](https://www.tensorflow.org/probability/api_docs/python/tfp/substrates/jax/mcmc/sample_chain) for it. It seems to me that it just does the same job here.